src/Message/Response.php

```diff
@@ -396,4 +405,109 @@ public function getRequestId()
 
         return null;
     }
+
+    /**
+     * Get the source reference
+     *
+     * @return null
+     */
+    public function getSourceReference()
+    {
+        if (isset($this->data['object']) && 'source' === $this->data['object']) {
+            return $this->data['id'];
+        }
+
+        return null;
+    }
+
+    /**
+     * @return bool
+     */
+    public function isRedirect()
+    {
+        if (isset($this->data['object']) && 'source' === $this->data['object']) {
+            if ($this->cardCan3DS() || ($this->isThreeDSecureSourcePending() && $this->getRedirectUrl() !== null)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Check if card requires 3DS
+     *
+     * @return bool
+     */
+    protected function cardCan3DS()
+    {
+        if (isset($this->data['type']) && 'card' === $this->data['type']) {
+            if (isset($this->data['card']['three_d_secure']) && ('required' === $this->data['card']['three_d_secure'] || 'optional' === $this->data['card']['three_d_secure'])) {
```

Comment:
https://stripe.com/docs/upgrades#2018-02-06
```
2018-02-06

    For a Source’s card[three_d_secure] property, adds "recommended" as a possible value.
    Previously, the only valid values were "not supported", "optional", and "required".
```
So you should also support here "recommended" value.
As it will be 3 possible values now then better to use "\in_array" function instead of 3 OR's :)
```
    protected function cardCan3DS()
    {
        if (isset($this->data['type']) && 'card' === $this->data['type']) {
            if (isset($this->data['card']['three_d_secure']) && \in_array($this->data['card']['three_d_secure'], ['required', 'optional', 'recommended'], true)) {
                return true;
            }
        }
        return false;
    }
```

One note though: im not sure if you want to always force "3d secure" if the value is "optional" here.

Reply:
@domis I actually coded this long time ago, and didn't realise that there's an update to 3DSecure of Stripe. Will update this soon.

Thanks, Good Catch!

Reply:
@domis86 I pushed a fix.

For your concern in "optional" value, i would recommend to enforce a 3DS for security purposes, it will help prevent fraudulent payments. 

We can have this as an improvement in the future if necessary, and I'll be happy to apply the changes, and everyone is welcome to send PR in case it fits for their organization's needs

Reply:
Ok - maybe is better safer to force the 3d Secure on "optional" indeed.